R/loo_compare.R

```diff
@@ -101,54 +126,122 @@ loo_compare.default <- function(x, ...) {
     loos <- x
   }
 
-  # If subsampling is used
+  # if subsampling is used
   if (any(sapply(loos, inherits, "psis_loo_ss"))) {
     return(loo_compare.psis_loo_ss_list(loos))
   }
 
+  # run pre-comparison checks
   loo_compare_checks(loos)
 
+  # compute elpd_diff and se_elpd_diff relative to best model
   comp <- loo_compare_matrix(loos)
   ord <- loo_compare_order(loos)
-
-  # compute elpd_diff and se_elpd_diff relative to best model
   rnms <- rownames(comp)
   diffs <- mapply(FUN = elpd_diffs, loos[ord[1]], loos[ord])
+  colnames(diffs) <- rnms
   elpd_diff <- apply(diffs, 2, sum)
   se_diff <- apply(diffs, 2, se_elpd_diff)
-  comp <- cbind(elpd_diff = elpd_diff, se_diff = se_diff, comp)
-  rownames(comp) <- rnms
 
-  # run order statistics-based checks on models
+  # compute probabilities that a model has worse elpd than the best model
+  # using a normal approximation (Sivula et al., 2025)
+  p_worse <- stats::pnorm(0, elpd_diff, se_diff)
+  p_worse[elpd_diff == 0] <- NA
+
+  # diagnostics to assess whether the normal approximation can be trusted
+  # * N < 100: small data (Sivula et al., 2025)
+  # * |elpd_diff| < 4: similar predictions (Sivula et al., 2025)
+  # * khat_diff > 0.5: possible outliers in differences (Sivula et al., 2025; Vehtari et al., 2024)
+  N <- nrow(diffs)
+  if (N < 100) {
+    diag_diff <- rep("N < 100", length(elpd_diff))
+    diag_diff[elpd_diff == 0] <- ""
+  } else {
+    diag_diff <- rep("", length(elpd_diff))
+    diag_diff[elpd_diff > -4 & elpd_diff != 0] <- "|elpd_diff| < 4"
+    khat_diff <- rep(NA, length(elpd_diff))
+    khat_diff[elpd_diff != 0] <- apply(
+      diffs[, elpd_diff != 0, drop = FALSE], 2,
+      function(x) ifelse(length(unique(x)) <= 20, NA, posterior::pareto_khat(x, tail = "both")
+    ))
+    diag_diff[khat_diff > 0.5] <- "khat_diff > 0.5"
+  }
+
+  # get khats for PSIS
+  khat_psis <- sapply(loos[ord],
+                      \(loo) {
```

Comment:
```suggestion
                      function(loo) {
```

This syntax is only available in more recent versions of R (I think >= 4.1), but loo can work with older R versions so we shouldn't use the new syntax unless we want to restrict loo to R>=4.1. If we want to do that we need to increase the R version listed in the DESCRIPTION file.